pkg/components/pluggable/discovery.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// AddServiceDiscoveryCallback adds a callback function that should be called when the given service was discovered.
+func AddServiceDiscoveryCallback(serviceName string, callbackFunc func(name string, dialer GRPCConnectionDialer)) {
+	onServiceDiscovered[serviceName] = callbackFunc
+}
+
+// removeExt removes file extension
+func removeExt(fileName string) string {
+	return fileName[:len(fileName)-len(filepath.Ext(fileName))]
+}
+
+const (
+	SocketFolderEnvVar  = "DAPR_COMPONENTS_SOCKETS_FOLDER"
+	defaultSocketFolder = "/tmp/dapr-components-sockets"
+)
+
+// GetSocketFolderPath returns the shared unix domain socket folder path
+func GetSocketFolderPath() string {
+	return utils.GetEnvOrElse(SocketFolderEnvVar, defaultSocketFolder)
+}
+
+type service struct {
+	// protoRef is the proto service name
+	protoRef string
+	// componentName is the component name that implements such service.
+	componentName string
+	// dialer is the used grpc connectiondialer.
+	dialer GRPCConnectionDialer
+}
+
+type reflectServiceClient interface {
+	ListServices() ([]string, error)
+}
+
+// serviceDiscovery returns all available discovered pluggable components services.
+// uses gRPC reflection package to list implemented services.
+func serviceDiscovery(reflectClientFactory func(string) (reflectServiceClient, func(), error)) ([]service, error) {
+	services := []service{}
+	componentsSocketPath := GetSocketFolderPath()
+	_, err := os.Stat(componentsSocketPath)
+
+	if os.IsNotExist(err) { // not exists is the same as empty.
+		return services, nil
+	}
+
+	log.Debugf("loading pluggable components under path %s", componentsSocketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := os.ReadDir(componentsSocketPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list pluggable components unix sockets")
+	}
+
+	for _, dirEntry := range files {
+		if dirEntry.IsDir() { // skip dirs
+			continue
+		}
+
+		f, err := dirEntry.Info()
+		if err != nil {
+			return nil, err
+		}
+
+		socket := filepath.Join(componentsSocketPath, f.Name())
+		if !utils.IsSocket(f) {
+			discoveryLog.Warnf("could not use socket for file %s", socket)
+			continue
+		}
+
+		refctClient, cleanup, err := reflectClientFactory(socket)
+		if err != nil {
+			return nil, err
+		}
+		defer cleanup()
+
+		serviceList, err := refctClient.ListServices()
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to list services")
+		}
+		dialer := socketDialer(socket)
+
+		componentName := removeExt(f.Name())
+		for _, svc := range serviceList {
+			services = append(services, service{
+				componentName: componentName,
+				protoRef:      svc,
+				dialer:        dialer,
+			})
+		}
+	}
+	log.Debugf("found %d pluggable component services", len(services)-1) // reflection api doesn't count.
+	return services, nil
+}
+
+// callback invoke callback function for each given service
+func callback(services []service) {
```

Comment:
Can we change this to use channels? This is Go, not Javascript ;-)

Reply:
I would be OK if this refactoring was a follow up PR in 1.10 so we can merge this in 1.9